packages/account/imAccount/factory.ts

```diff
@@ -0,0 +1,151 @@
+  private entryPointAddress: string
+  private fallbackHandlerAddress: string
+  private validator: Validator
+
+  public constructor(opts: {
+    factoryAddress: string
+    implementationAddress: string
+    entryPointAddress: string
+    validator: Validator
+    fallbackHandlerAddress: string
+    salt: BigNumberish
+    nodeRpcUrl: string
+  }) {
+    this.node = new ethers.JsonRpcProvider(opts.nodeRpcUrl)
+    this.factory = new ethers.Contract(
+      opts.factoryAddress,
+      imAccountFactoryMetadata.abi,
+      this.node
+    )
+    this.factoryOwner = new ethers.Wallet(config.account.operator.privateKey)
+    this.imAccount = new ethers.Contract(
+      opts.implementationAddress,
+      imAccountMetadata.abi,
+      this.node
+    )
+    this.entryPointAddress = opts.entryPointAddress
+    this.validator = opts.validator
+    this.fallbackHandlerAddress = opts.fallbackHandlerAddress
+    this.salt = opts.salt
+  }
+
+  public async getAddress() {
+    return ethers.zeroPadValue(
+      ethers.stripZerosLeft(
+        // The name of `getAddress` conflicts with the function on ethers.Contract.
+        // So we build call data from interface and directly send through node rpc provider.
+        await this.node.call(
+          await this.factory
+            .getFunction("getAddress")
+            .populateTransaction(this.salt)
+        )
+      ),
+      20
+    )
+  }
+
+  public async getInitCode() {
+    const initializer = await this.getInitializer()
+    const createAccountRequest = {
+      initializer: initializer,
+      salt: this.salt
+    }
+    const signature = await this.signCreateAccount(createAccountRequest)
+    console.log("signature: ", signature)
+
+    const params = this.factory.interface.encodeFunctionData(
+      "createAccountWithSignature",
+      [createAccountRequest, signature]
+    )
+    return ethers.concat([await this.factory.getAddress(), params])
+  }
+
+  private async getInitializer() {
+    const ownerValidatorInitData =
+      await this.validator.getOwnerValidatorInitData()
+    const { data } = await this.imAccount
+      .getFunction("initialize")
+      .populateTransaction(
+        this.entryPointAddress,
+        this.fallbackHandlerAddress,
+        this.validator.getAddress(),
+        ownerValidatorInitData
+      )
+    return data
+  }
+
+  // Currently, the factory owner's key is imported from `~/config/test` to sign the request for creating an account.
+  // However, in future developments, we will require a factory signing server.
+  private async signCreateAccount(createAccountRequest: CreateAccount) {
+    const EIP712_DOMAIN_HASH = ethers.keccak256(
+      ethers.toUtf8Bytes(
+        "EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"
+      )
+    )
+    const nameHash = ethers.keccak256(ethers.toUtf8Bytes("imAccountFactory"))
+    const versionHash = ethers.keccak256(ethers.toUtf8Bytes("v0"))
+    const chainId = this.node._network.chainId
```

Comment:
nit: It could be better to depend on public interface

```suggestion
    const { chainId }  = await this.node.getNetwork()
```